Create an EQC statem test for the DVV style riak_object

russelldb opened the pull request:
When making test I had to cut and paste the `riak_kv_vnode:put_merge` into the test. The same code is also cut and pasted into `ec_eqc`. Seems an obvious candidate for a refactor. And `riak_object` seemed like the location since `put_merge` tinkered with `riak_object` state.

russelldb committed:
WIP a statem eqc test  for ro:merge that uses DVVSet for a model

russelldb committed:
Add copious comments and a command for stale puts

russelldb committed:
Move put merge code inside riak_object

Since the put_merge code was copied into two tests, both of which
only manipulate riak_object, it seemed a refactor made sense.

Looking at the put merge code, it only manipulates internal state
of riak_object, so I think it is the correct place for the code.

It also smooths the way for a general refactor of riak_object
into a CRDT and riak_kv into riak_crdt (mwah ha ha ha ha!)

russelldb committed:
Slight doc / comment clean up

russelldb committed:
Remove comments about copied code, refactored the copied code out

seancribbs commented on src/riak_object.erl:
```diff
@@ -708,6 +695,38 @@ is_updated(_Object=#r_object{updatemetadata=M,updatevalue=V}) ->
+            FrontierClock = vclock:increment(Actor, Timestamp, MergedClock),
+            {ok, Dot} = vclock:get_entry(Actor, FrontierClock),
+            %% Assign an event to the new value
+            DottedPutObject = assign_dot(NewObject, Dot, dvv_enabled()),
```
Didn't you just remove this `assign_dot` function from the module?

russelldb replied:
Removed the exported 'assign_dot/2', kept this private one.

seancribbs replied:
Yes, but it's missing from the [diff above](https://github.com/basho/riak_kv/pull/807/files#diff-23f6ea5339469f5d292dca9ba27d2914L586).

seancribbs replied:
Oh n/m disregard

russelldb committed:
eqc_group_commands style

russelldb commented on test/riak_object_dvv_statem.erl:
```diff
@@ -0,0 +1,375 @@
+    ].
+
+%% Like when a user does a put without first fetching (blank vv) @see
+%% next_state/3 for the return tuple being prepended to vnode_data
```
note to me, when I address Dave's review comments, fix this too.

bookshelfdave commented on test/riak_object_dvv_statem_group.erl~:
```diff
@@ -0,0 +1,298 @@
+%% -------------------------------------------------------------------
```
look like an emacs backup file was added, can you remove?

bookshelfdave commented on test/riak_object_dvv_statem.erl:
```diff
@@ -0,0 +1,375 @@
+    N > 0 andalso length(VNodes) < N * 2.
+
+make_ring_args(#state{vnodes=VNodes, n=N}) ->
+    [VNodes, vector(N, binary(8))].
```
if I'm reading this right, Vnodes is ignored in make_ring_next, why bother returning a list here?

bookshelfdave replied:
my bad, _args() needs to return [term()]. But VNodes is ignored.

bookshelfdave commented on test/riak_object_dvv_statem.erl:
```diff
@@ -0,0 +1,375 @@
+%% @see put/4, put/5, replicate/4 and update/7 for how data gets
+%% stored
+get(VNode1, R, VNodes, VNodeData) ->
+    Start = index_of(VNode1, VNodes),
```
redacted comment regarding index_of/2 returning `undefined`. VNodes won't be null because of make_ring_pre

bookshelfdave commented on test/riak_object_dvv_statem.erl:
```diff
@@ -0,0 +1,375 @@
+
+%% Generate a list of vnodes, RW long, starting with the vnode at
+%% index Start
+preflist(Start, RW, VNodes) ->
```
it almost seems like preflist/\* could have it's own EQC prop to verify it's working

russelldb replied:
Yeah, maybe it just needs to pick any R vnodes are random from the list, that is all it should do really.

bookshelfdave commented on test/riak_object_dvv_statem.erl:
```diff
@@ -0,0 +1,375 @@
+%% next_state/3 for the return tuple being prepended to vnode_data
+put(Value, Coord, VNodeData, Time) ->
+    {RObj, DVV} = coord_put(Coord, Value, Time, VNodeData),
+    {Coord, RObj, DVV, Time}.
```
can you verify that this tuple is correct? Seems like the first position should be Value, as it's placed into VNodeData `[{binary(), riak_object:object(), dvvset:dvvset(), pos_integer()}],`

russelldb replied:
Vnode data is a four tuple of `{VNodeId :: binary(), riak_object:object(), dvvset:dvvset(), Time :: pos_integer()}`

Value is _inside_ riak_object and dvvset.

bookshelfdave replied:
ah, can you clarify that in the record definition?

russelldb replied:
Will do, yes.

bookshelfdave commented on test/riak_object_dvv_statem.erl:
```diff
@@ -0,0 +1,375 @@
+replicate_args(#state{vnodes=VNodes, vnode_data=VNodeData, time=Time}) ->
+    [
+     elements(VNodes), %% Replicate from
+     elements(VNodes), %% replicate too
```
do you care if From == To?

russelldb replied:
I guess I'd care if it From was _always_ To, yes.

bookshelfdave commented:
this test seems fine. It passed against 100k EQC runs.

bookshelfdave commented:
I noticed src/riak_kv_mutator.erl is still in your branch, can you rebase against develop?

russelldb commented:
I addressed all comments except that desire to simplify the fake preflist generation. I really _should_ address that, but I want the code this test tests into the build on Friday. Honestly I'll simplify the test in a later PR, since it will need to change to address that weird forwarding siblings issue.

I rebased, and now when I try and push git tells me I need to pull and merge. Do I force push? Some other thing? Please can you help?

bookshelfdave commented:
I'd say create a new rebased branch + PR, close this one but reference it from the new PR.

russelldb commented:
Closing and opening new PR for rebased branch.

bookshelfdave commented:
Link to new PR: https://github.com/basho/riak_kv/pull/835
